From 848dfc1c3f23eea5b6035e42fc0d77d39e5b4528 Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:19:49 -0400
Subject: [net] netfilter: nf_tables: fix error return code

Message-id: <20141021161949.22888.15306.stgit@localhost.localdomain>
Patchwork-id: 97986
O-Subject: [RHEL7 net PATCH 089/124] netfilter: nf_tables: fix error return code
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 609ccf087747de48ef52160f93e0df864c532a61

 commit 609ccf087747de48ef52160f93e0df864c532a61
 Author: Julia Lawall <Julia.Lawall@lip6.fr>
 Date:   Thu Aug 7 14:49:08 2014 +0200

    netfilter: nf_tables: fix error return code

    Convert a zero return value on error to a negative one, as returned
    elsewhere in the function.

    A simplified version of the semantic match that finds this problem is as
    follows: (http://coccinelle.lip6.fr/)

    // <smpl>
    @@
    identifier ret; expression e1,e2;
    @@
    (
    if (\(ret < 0\|ret != 0\))
     { ... return ret; }
    |
    ret = 0
    )
    ... when != ret = e1
        when != &ret
    *if(...)
    {
      ... when != ret = e2
          when forall
     return ret;
    }
    // </smpl>

    Signed-off-by: Julia Lawall <Julia.Lawall@lip6.fr>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/netfilter/nf_tables_api.c b/net/netfilter/nf_tables_api.c
index b912641..12ea8fc 100644
--- a/net/netfilter/nf_tables_api.c
+++ b/net/netfilter/nf_tables_api.c
@@ -3137,8 +3137,10 @@ static int nft_del_setelem(struct nft_ctx *ctx, struct nft_set *set,
   goto err2;
 
  trans = nft_trans_elem_alloc(ctx, NFT_MSG_DELSETELEM, set);
- if (trans == NULL)
+ if (trans == NULL) {
+  err = -ENOMEM;
   goto err2;
+ }
 
  nft_trans_elem(trans) = elem;
  list_add_tail(&trans->list, &ctx->net->nft.commit_list);
-- 
1.7.1